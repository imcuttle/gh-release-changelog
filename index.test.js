const { ghReleaseChangelog, ghReleaseChangelogMonorepo } = require("./");
const utils = require("./utils");
const path = require("path");
const { testRequest, testGetTags } = require("@actions/github");

process.env.GITHUB_TOKEN = "noop";

const fixture = (name) => path.resolve(__dirname, "fixture", name);

jest.mock("@actions/github", () => {
  const testRequest = jest.fn(() => {});
  const testGetTags = jest.fn(() => {});

  return {
    testGetTags,
    testRequest,
    getOctokit: () => {
      return {
        request: (url, params) => {
          testRequest(params);
          if (url === "GET /repos/{owner}/{repo}/git/matching-refs/{ref}") {
            return testGetTags(params);
          }
          return;
        },
      };
    },
  };
});

beforeEach(() => {
  testRequest.mockClear();
  testGetTags.mockClear();
});

describe("ghReleaseChangelog", () => {
  test("valid", async () => {
    await ghReleaseChangelog({
      cwd: fixture("valid"),
      tag: "2.0.0",
      githubToken: "noop",
      label: "@rcp/abc",
      skipEnvGithubRepoInfer: true,
    });

    expect(testRequest).toHaveBeenCalledTimes(2);
    expect(testRequest.mock.calls).toMatchSnapshot();
  });
  test("valid-long v1.0.2", async () => {
    await ghReleaseChangelog({
      cwd: fixture("valid-long"),
      tag: "v1.0.2",
      githubToken: "noop",
      skipEnvGithubRepoInfer: true,
      repoName: "gh-release-changelog",
      repoOwner: "imcuttle",
      checkStandardVersion: true,
      checkPkgAvailable: true,
    });

    expect(testRequest.mock.calls).toMatchSnapshot();
    expect(testRequest).toHaveBeenCalledTimes(1);
  });

  test("valid-long v1.0.2-beta.3", async () => {
    await ghReleaseChangelog({
      cwd: fixture("valid-long"),
      tag: "v1.0.2-beta.3",
      githubToken: "noop",
      skipEnvGithubRepoInfer: true,
      repoName: "gh-release-changelog",
      repoOwner: "imcuttle",
      checkStandardVersion: false,
      checkPkgAvailable: true,
      skipFromTagGitInfer: true
    });

    expect(testRequest.mock.calls).toMatchSnapshot();
    expect(testRequest).toHaveBeenCalledTimes(2);
  });

  test("lerna-sub not-found", async () => {
    await ghReleaseChangelog({
      cwd: fixture("lerna-sub"),
      tag: "v1.0.0",
      githubToken: "noop",
      label: "@rcp/abc",
      skipEnvGithubRepoInfer: true,
    });
    expect(testRequest).toHaveBeenCalledTimes(0);
  });

  test("lerna-sub version bump only", async () => {
    await ghReleaseChangelog({
      cwd: fixture("lerna-sub"),
      tag: "v1.0.1",
      githubToken: "noop",
      label: "@rcp/abc",
      skipEnvGithubRepoInfer: true,
    });
    expect(testRequest).toHaveBeenCalledTimes(0);
  });
});

describe("ghReleaseChangelogMonorepo", () => {
  it("lerna", async function () {
    await ghReleaseChangelogMonorepo({
      cwd: fixture("monorepo/lerna"),
      tag: "1.0.0",
      repoName: "a",
      repoOwner: "o",
    });
    expect(testRequest).toHaveBeenCalledTimes(1);
    expect(testRequest.mock.calls[0]).toMatchSnapshot();
  });

  it("pnpm", async function () {
    testGetTags.mockReturnValueOnce({
      data: [
        {
          ref: "refs/tags/a@1.0.0",
        },
        {
          ref: "refs/tags/a@1.0.0-0",
        },
      ],
    });
    await ghReleaseChangelogMonorepo({
      cwd: fixture("monorepo/pnpm"),
      repoName: "a",
      repoOwner: "o",
      tag: "a@1.0.0",
    });

    expect(testRequest).toHaveBeenCalledTimes(2);
    expect(testRequest.mock.calls).toMatchSnapshot();
  });

  it("workspaces", async function () {
    await ghReleaseChangelogMonorepo({
      cwd: fixture("monorepo/workspaces"),
      repoName: "a",
      repoOwner: "o",
      tag: "1.0.0",
    });
    expect(testRequest).toHaveBeenCalledTimes(1);
    expect(testRequest.mock.calls[0]).toMatchSnapshot();
  });

  it("same-version-root-changelog", async function () {
    await ghReleaseChangelogMonorepo({
      cwd: fixture("monorepo/same-version-root-changelog"),
      tag: "1.0.0",
      repoName: "a",
      repoOwner: "o",
    });
    expect(testRequest).toHaveBeenCalledTimes(1);
    expect(testRequest.mock.calls[0]).toMatchSnapshot();
  });
});

describe("utils", () => {
  it("parserVersion", function () {
    expect(utils.parserVersion("v1.0.2-beta.2")).toMatchInlineSnapshot(`
      Object {
        "version": "1.0.2-beta.2",
      }
    `);
    expect(utils.parserVersion("@rcp/asdsa@1.0.2-beta.2"))
      .toMatchInlineSnapshot(`
      Object {
        "name": "@rcp/asdsa",
        "version": "1.0.2-beta.2",
      }
    `);

    expect(utils.parserVersion("@rcp/asdsa@1.0.2-beta.2 123"))
      .toMatchInlineSnapshot(`
      Object {
        "name": "@rcp/asdsa",
        "version": "1.0.2-beta.2",
      }
    `);
  });
});
