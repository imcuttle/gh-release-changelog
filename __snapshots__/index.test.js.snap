// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ghReleaseChangelog lerna-sub not-found 1`] = `
Object {
  "depth": undefined,
  "fromTag": undefined,
  "githubToken": "noop",
  "label": "@rcp/abc",
  "releaseNote": "",
  "repoName": "rcp.git",
  "repoOwner": "imcuttle",
  "tag": "v1.0.0",
}
`;

exports[`ghReleaseChangelog lerna-sub version bump only 1`] = `
Object {
  "depth": 1,
  "fromTag": undefined,
  "githubToken": "noop",
  "label": "@rcp/abc",
  "releaseNote": "",
  "repoName": "rcp.git",
  "repoOwner": "imcuttle",
  "tag": "v1.0.1",
}
`;

exports[`ghReleaseChangelog valid 1`] = `
Object {
  "depth": 2,
  "fromTag": undefined,
  "githubToken": "noop",
  "label": "@rcp/abc",
  "releaseNote": "## @rcp/abc

### Bug Fixes

*   should remove container firstly ([4e0d087](https://github.com/imcuttle/rcp/commit/4e0d087))
*   should remove container firstly ([8dfcc84](https://github.com/imcuttle/rcp/commit/8dfcc84))
*   use extendDictionary ([330edf0](https://github.com/imcuttle/rcp/commit/330edf0))
*   use.fetcher with key ([fe45ac1](https://github.com/imcuttle/rcp/commit/fe45ac1))
*   use.i18ncontext ([da70207](https://github.com/imcuttle/rcp/commit/da70207))

### Features

*   add c.preventfastop ([44bf84f](https://github.com/imcuttle/rcp/commit/44bf84f))
*   add eq option ([631dfe7](https://github.com/imcuttle/rcp/commit/631dfe7))
*   add hook ([300eb94](https://github.com/imcuttle/rcp/commit/300eb94))
*   add use.i18n ([8f2b256](https://github.com/imcuttle/rcp/commit/8f2b256))
*   allow customized eq function ([79e92ce](https://github.com/imcuttle/rcp/commit/79e92ce))
*   remove global option ([55f611a](https://github.com/imcuttle/rcp/commit/55f611a))
*   use es6 instead of es5 ([3dcc2a8](https://github.com/imcuttle/rcp/commit/3dcc2a8))
*   util.createlogger log alias to info ([8757ae8](https://github.com/imcuttle/rcp/commit/8757ae8))

### Performance Improvements

*   set defaultLanguage to defaultProps ([cf01859](https://github.com/imcuttle/rcp/commit/cf01859))

### BREAKING CHANGES

*   use es6 instead of es5
*   remove global option

Full Changelog: [v1.0.0](https://github.com/imcuttle/rcp/commits/v1.0.0)",
  "repoName": "rcp",
  "repoOwner": "imcuttle",
  "tag": "v1.0.0",
}
`;
